File: windows/PetRunner.Core/DefaultPet.cs
namespace PetRunner.Core;

public static class DefaultPet
{
    public const string Id = "maomao";
    public const string DownloadPetsUrl = "https://pet-runner.com";
    public const string BundleRootRelativePath = "DefaultPets";
    public const string BundleRelativePath = "DefaultPets/maomao";
}

public static class PetSelectionOrdering
{
    public static IReadOnlyList<PetDescriptor> OrderedCandidates(IReadOnlyList<PetDescriptor> pets, string? selectedId)
    {
        if (pets.Count == 0) return pets;
        var preferred =
            pets.FirstOrDefault(pet => string.Equals(pet.Id, selectedId, StringComparison.Ordinal))
            ?? pets.FirstOrDefault(pet => string.Equals(pet.Id, DefaultPet.Id, StringComparison.Ordinal))
            ?? pets[0];
        return [preferred, .. pets.Where(pet => !string.Equals(pet.Id, preferred.Id, StringComparison.Ordinal))];
    }
}

public sealed class DefaultPetInstaller
{
    /// <summary>
    /// Installs every valid pet package under <paramref name="bundledRoot"/> that is missing
    /// from the library. Never overwrites an existing package. Invalid folders are skipped.
    /// </summary>
    public int InstallAllMissing(string bundledRoot, string petsDirectory)
    {
        if (!Directory.Exists(bundledRoot)) return 0;
        var installed = 0;
        foreach (var package in Directory.EnumerateDirectories(bundledRoot).OrderBy(path => path, StringComparer.Ordinal))
        {
            try
            {
                if (InstallIfMissing(package, petsDirectory)) installed++;
            }
            catch (Exception)
            {
                // Skip invalid bundled folders so one bad package does not block the rest.
            }
        }
        return installed;
    }

    /// <summary>
    /// Copies a bundled pet package into the library when missing. Never overwrites.
    /// Destination folder uses the package id from pet.json.
    /// </summary>
    public bool InstallIfMissing(string bundledPackage, string petsDirectory)
    {
        if (!Directory.Exists(bundledPackage)) return false;

        var pet = PetPackageLoader.LoadPackage(bundledPackage);
        var id = RequireSafePetId(pet.Id);
        var petsRoot = Path.GetFullPath(petsDirectory);
        var destination = Path.GetFullPath(Path.Combine(petsRoot, id));
        EnsureDestinationInsidePetsRoot(destination, petsRoot);
        if (Directory.Exists(destination)) return false;

        Directory.CreateDirectory(petsRoot);
        CopyDirectory(bundledPackage, destination);
        _ = PetPackageLoader.LoadPackage(destination);
        return true;
    }

    internal static string RequireSafePetId(string id)
    {
        var trimmed = id.Trim();
        if (trimmed.Length == 0
            || trimmed is "." or ".."
            || trimmed.Contains('/')
            || trimmed.Contains('\\')
            || Path.IsPathRooted(trimmed))
        {
            throw new InvalidDataException($"pet id {id} is not a safe library folder name");
        }
        return trimmed;
    }

    private static void EnsureDestinationInsidePetsRoot(string destination, string petsRoot)
    {
        var root = petsRoot.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
            + Path.DirectorySeparatorChar;
        if (!destination.StartsWith(root, StringComparison.OrdinalIgnoreCase)
            && !string.Equals(destination, petsRoot, StringComparison.OrdinalIgnoreCase))
        {
            throw new InvalidDataException("pet install destination escapes the pets directory");
        }
    }

    private static void CopyDirectory(string source, string destination)
    {
        Directory.CreateDirectory(destination);
        foreach (var file in Directory.EnumerateFiles(source))
            File.Copy(file, Path.Combine(destination, Path.GetFileName(file)));
        foreach (var directory in Directory.EnumerateDirectories(source))
            CopyDirectory(directory, Path.Combine(destination, Path.GetFileName(directory)));
    }
}

File: Sources/PetRunnerCore/DefaultPet.swift
import Foundation

/// Built-in default pet shipped with PetRunner and preferred on first launch.
public enum DefaultPet {
    public static let id = "maomao"
    public static let downloadPetsURL = URL(string: "https://pet-runner.com")!
    public static let bundleRootRelativePath = "DefaultPets"
    public static let bundleRelativePath = "DefaultPets/maomao"
}

public enum PetSelectionOrdering {
    /// Prefer an explicitly selected pet, then the built-in default id, then scan order.
    public static func orderedCandidates(from pets: [PetDescriptor], selectedID: String?) -> [PetDescriptor] {
        guard !pets.isEmpty else { return [] }
        let preferred =
            pets.first { $0.id == selectedID }
            ?? pets.first { $0.id == DefaultPet.id }
            ?? pets[0]
        return [preferred] + pets.filter { $0.id != preferred.id }
    }
}

public enum DefaultPetInstallError: Error, Equatable {
    case unsafePetID(String)
    case destinationEscapesPetsDirectory
}

public struct DefaultPetInstaller {
    private let fileManager: FileManager
    private let loader: PetPackageLoader

    public init(fileManager: FileManager = .default, loader: PetPackageLoader = .init()) {
        self.fileManager = fileManager
        self.loader = loader
    }

    /// Installs every valid pet package under `bundledRoot` that is missing from the library.
    /// Never overwrites an existing package. Invalid folders are skipped.
    @discardableResult
    public func installAllMissing(bundledRoot: URL, into petsDirectory: URL) throws -> Int {
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: bundledRoot.path, isDirectory: &isDirectory), isDirectory.boolValue else {
            return 0
        }
        let packages = try fileManager.contentsOfDirectory(
            at: bundledRoot,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        )
        .filter { url in
            (try? url.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true
        }
        .sorted { $0.lastPathComponent < $1.lastPathComponent }

        var installed = 0
        for package in packages {
            do {
                if try installIfMissing(bundledPackage: package, into: petsDirectory) {
                    installed += 1
                }
            } catch {
                // Skip invalid bundled folders so one bad package does not block the rest.
            }
        }
        return installed
    }

    /// Copies a bundled pet package into the library when missing. Never overwrites an existing package.
    /// Destination folder uses the package id from pet.json.
    @discardableResult
    public func installIfMissing(bundledPackage: URL, into petsDirectory: URL) throws -> Bool {
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: bundledPackage.path, isDirectory: &isDirectory), isDirectory.boolValue else {
            return false
        }

        let pet = try loader.loadPackage(at: bundledPackage)
        let id = try Self.requireSafePetID(pet.id)
        let petsRoot = petsDirectory.standardizedFileURL
        let destination = petsRoot.appendingPathComponent(id, isDirectory: true).standardizedFileURL
        try Self.ensureDestinationInsidePetsRoot(destination, petsRoot: petsRoot)
        if fileManager.fileExists(atPath: destination.path) { return false }

        try fileManager.createDirectory(at: petsRoot, withIntermediateDirectories: true)
        try fileManager.copyItem(at: bundledPackage, to: destination)
        _ = try loader.loadPackage(at: destination)
        return true
    }

    static func requireSafePetID(_ id: String) throws -> String {
        let trimmed = id.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty
            || trimmed == "."
            || trimmed == ".."
            || trimmed.contains("/")
            || trimmed.contains("\\")
            || (trimmed as NSString).isAbsolutePath
        {
            throw DefaultPetInstallError.unsafePetID(id)
        }
        return trimmed
    }

    private static func ensureDestinationInsidePetsRoot(_ destination: URL, petsRoot: URL) throws {
        let rootPath = petsRoot.path
        let destinationPath = destination.path
        let prefix = rootPath.hasSuffix("/") ? rootPath : rootPath + "/"
        guard destinationPath == rootPath || destinationPath.hasPrefix(prefix) else {
            throw DefaultPetInstallError.destinationEscapesPetsDirectory
        }
    }
}
